### DISCUSSION TITLE
Separate cache hash length and folder depth

### DISCUSSION THREAD

**Discussion by ronaldbarendse** [SixLabors/ImageSharp.Web#210] [Ideas] [answered]
Currently, the length of the created cache hash can be configured on the middleware options:
https://github.com/SixLabors/ImageSharp.Web/blob/a411d0e188fc6768db59d25697dd92423159806b/src/ImageSharp.Web/Middleware/ImageSharpMiddlewareOptions.cs#L87-L91

And this is used as a parameter on `ICacheHash.Create()` (so it makes sense that this isn't a setting specific to the `ICacheHash` implementation and a more general one defined in `ImageSharpMiddlewareOptions`):
https://github.com/SixLabors/ImageSharp.Web/blob/c84f4ba28253170d67c54cb98cbfc18c2d5c163b/src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs#L264

But this setting is also used by `PhysicalFileSystemCache` to specify the depth of the nested cache folders:
https://github.com/SixLabors/ImageSharp.Web/blob/af70a40d68d4edb025dba99b99e2473194f267ba/src/ImageSharp.Web/Caching/PhysicalFileSystemCache.cs#L83

So if you currently want to increase the cache hash length (e.g. to reduce collisions), it will also increase the cache folder depth. And because this is linked, it will only ever have a single image and metadata file inside a folder. I recon the folder nesting is done to prevent enumerating large amount of files in a single directory, but the additional nested folders might actually reduce I/O performance (because each folder still requires something to be written into the MFT record).

I would suggest the following changes:
- Change setting name from `CachedNameLength` to `CacheHashLength`, since is a more suitable name for where it's actually used and the `IImageCache` is otherwise responsible for creating the final file name;
- Rename the default `CacheHash` implementation to `SHA256CacheHash`, to make it more explicit how the hash is generated;
- Add a new `CacheFolderDepth` setting to `PhysicalFileSystemCacheOptions`, so this can be configured separately;
- Optionally: because the nested folders already contain the first _n_ characters from the file name, these can be stripped from the final name, e.g. `027ee1e815be.meta` (12 characters) could be stored as `0/2/7/e/e/1/e/8/15be.meta` (with a depth of 8), instead of `0/2/7/e/e/1/e/8/027ee1e815be.meta`. This decreases the path length, so you're also less likely to hit maximum path length limits...

I've already created a [branch with the first 3 changes](https://github.com/SixLabors/ImageSharp.Web/compare/master...ronaldbarendse:feature/cachehash-cleanup?expand=1), which I can submit as PR if you'd like.

**Comment by JimBobSquarePants** [comment, marked as answer]
<strike>Dude. I’m not changing the cache. </strike>

Apologies... I need to stop reading these things first thing in the morning, I'm always grumpy.

If we make these changes we need to ensure that the default behavior is backwards compatible, we've got users out there with millions of cached files. I've already made changes to #206 to ensure that users can turn on the old behavior and keep their existing cache structure.

We have to do some sanity checking also to ensure that the depth cannot be greater than the hash.

>Optionally: because the nested folders already contain the first n characters from the file name, these can be stripped from the final name, e.g. `027ee1e815be.meta` (12 characters) could be stored as `0/2/7/e/e/1/e/8/15be.meta` (with a depth of 8), instead of `0/2/7/e/e/1/e/8/027ee1e815be.meta`. This decreases the path length, so you're also less likely to hit maximum path length limits...

This is an interesting suggestion. We'd have to think up a decent property name in the cache options -perhaps `TrimNestedCharacters`? (_suggestions welcome_) and default it to `false`, again for backwards compatibility. We'd also have to ensure that trimming leading characters is limited so that at least one char remains.

For information as why nesting occurs. Here's a feast of information on NTSF file/folder performance.
https://stackoverflow.com/questions/197162/ntfs-performance-and-large-volumes-of-files-and-directories

**Reply by ronaldbarendse** [reply]
Thanks for re-reading (and the apology) 😄

We don't need an additional property though, because the existing functionality will be used when both the `CacheHashLength` and `CacheFolderDepth` are set to the same value! I wouldn't use the same values as default though, as explained above.

The optimal value for the nested folder depth will depend on the following:
- The `CacheHashLength`, as the cache hash is used as file name and longer file names result in more files in the same directory (if the cache folder depth is kept the same);
- The characters returned from the `ICacheHash` implementation: currently the default hash is HEX-encoded, so that results in 16 different characters per position;
- The maximum amount of files you would (theoretically) want to have in a single folder.

If we keep the existing `CacheHashLength` value of 12 and the default hash implementation, it can theoretically store 16<sup>12</sup> files, which would result in the following amount of files/folder based on the `CacheFolderDepth`:

| CacheFolderDepth | Files/folder |
|-|-|
| 0 | 16<sup>12</sup> = 281 trillion |
| 1 | 16<sup>11</sup> = 17 trillion |
| 2 | 16<sup>10</sup> = 1 trillion |
| 3 | 16<sup>9</sup> = 68 billion |
| 4 | 16<sup>8</sup> = 4 billion |
| 5 | 16<sup>7</sup> = 268 million |
| 6 | 16<sup>6</sup> = 16 million |
| 7 | 16<sup>5</sup> = 1 million |
| 8 | 16<sup>4</sup> = 65.536 |
| 9 | 16<sup>3</sup> = 4096 |
| 10 | 16<sup>2</sup> = 256 |
| 11 | 16<sup>1</sup> = 16 |
| 12 | 16<sup>0</sup> = 1 |

Based on this, I think a default cache folder depth of 8 is therefore a nice balance between amount of nested folders and total files per folder (which only exist in the deepest folder).

**Reply by ronaldbarendse** [reply]
~We don't need an additional property though~ I need to re-read more as well 😇 Yes, we still need something to make this backwards compatible: I suggest adding a setting like `UseFullName` or `UseLegacyName` on `PhysicalFileSystemCacheOptions`.

**Reply by ronaldbarendse** [reply]
I've created the PR: https://github.com/SixLabors/ImageSharp.Web/pull/212

**Reply by ronaldbarendse** [reply]
And thanks for the link on NTSF file/folder performance 👍🏻 If we're using a default cache hash length of 12 and cache depth of 8, the resulting file name length would be 6 characters, so you would also solve any performance issues with short-file name generation (because the first 6 characters are always unique) 🎉

**Reply by ronaldbarendse** [reply]
I stand by my original statement: we don't need an additional property/setting: https://github.com/SixLabors/ImageSharp.Web/pull/213 😈

**Reply by hey-red** [reply]
It's also a nice option for linux based OS, because most of distributions uses ext4 as default filesystem.
So, in default installations we have limited number of inodes and cache takes one inode for every created directory. This can cause exhaustion of the pool of inodes and your application has stopped working.

As I know, with ext4 we cannot resize inode table(only way to recreate volume and specify number of inodes).
